File: exercism/ruby/space-age/space_age.rb
class SpaceAge
  attr_reader :seconds
  def initialize (seconds)
    @seconds = seconds
  end
  PLANETS = [
    ["earth", 1],
    ["mercury", 0.2408467],
    ["venus", 0.61519726],
    ["mars", 1.8808158],
    ["jupiter", 11.862615],
    ["saturn", 29.447498],
    ["uranus", 84.016846],
    ["neptune", 164.79132]
  ]

  PLANETS.each do |name, factor|
    define_method("on_"+ name) do
      (@seconds / 31557600.0 / factor).round(2)
    end
  end
end

File: exercism/ocaml/space-age/space_age.ml
open Core.Std

type planet = Mercury | Venus | Earth | Mars
            | Jupiter | Saturn | Neptune | Uranus

let seconds_in_earth_year: float = 365.25 *. 24.0 *. 60.0 *. 60.0

let relative_to_earth: planet -> float = function
  | Earth -> 1.0
  | Mercury -> 0.2408467
  | Venus -> 0.61519726
  | Mars -> 1.8808158
  | Jupiter -> 11.862615
  | Saturn -> 29.447498
  | Uranus -> 84.016846
  | Neptune -> 164.79132

let age_on (p: planet) (seconds: int) : float =
  float seconds /. seconds_in_earth_year /. relative_to_earth p
